pandas/_libs/src/vendored/numpy/datetime/np_datetime.c

```diff
@@ -752,66 +752,57 @@ void pandas_timedelta_to_timedeltastruct(npy_timedelta td,
     }
 
     const npy_int64 per_day = sec_per_day * per_sec;
-    npy_int64 frac;
+    const int sign = td < 0 ? -1 : 1;
     // put frac in seconds
-    if (td < 0 && td % per_sec != 0)
-      frac = td / per_sec - 1;
-    else
-      frac = td / per_sec;
+    npy_int64 sfrac = td / per_sec;
+    if (sign < 0) {
+      if (td % per_sec != 0)
+        sfrac -= 1;
 
-    const int sign = frac < 0 ? -1 : 1;
-    if (frac < 0) {
       // even fraction
-      if ((-frac % sec_per_day) != 0) {
-        out->days = -frac / sec_per_day + 1;
-        frac += sec_per_day * out->days;
+      if ((-sfrac % sec_per_day) != 0) {
+        out->days = sfrac / sec_per_day - 1;
+        sfrac -= sec_per_day * out->days;
       } else {
-        frac = -frac;
+        if (sfrac <= sec_per_day) {
```

Comment:
This seems to add more branching than what we had before, so I'm a little hesitant to say this is faster overall even though it may show up in some of our benchmarks. What kind of difference are you seeing in the current state?